lib/harmony/harmonytask/taskTypeHandler.go

```diff
@@ -0,0 +1,274 @@
+
+			h.TaskEngine.tryAllWork <- true // Activate tasks in this machine
+		}()
+
+		done, doErr = h.Do(*tID, func() bool {
+			var owner int
+			// Background here because we don't want GracefulRestart to block this save.
+			err := h.TaskEngine.db.QueryRow(context.Background(),
+				`SELECT owner_id FROM harmony_task WHERE id=$1`, *tID).Scan(&owner)
+			if err != nil {
+				logger.Error("Cannot determine ownership: ", err)
+				return false
+			}
+			return owner == h.TaskEngine.ownerID
+		})
+		if doErr != nil {
+			logger.Error("Do("+h.Name+", taskID="+strconv.Itoa(int(*tID))+") returned error: ", doErr)
+		}
+	}()
+	return true
+}
+
+func (h *taskTypeHandler) recordCompletion(tID TaskID, workStart time.Time, done bool, doErr error) {
+	workEnd := time.Now()
+
+	cm, err := h.TaskEngine.db.BeginTransaction(h.TaskEngine.ctx, func(tx *harmonydb.Tx) bool {
+		var postedTime time.Time
+		err := tx.QueryRow(`SELECT posted_time FROM harmony_task WHERE id=$1`, tID).Scan(&postedTime)
+		if err != nil {
+			logger.Error("Could not log completion: ", err)
+			return false
+		}
+		result := "unspecified error"
+		if done {
+			_, err = tx.Exec("DELETE FROM harmony_task WHERE id=$1", tID)
+			if err != nil {
+				logger.Error("Could not log completion: ", err)
+				return false
+			}
+			result = ""
+		} else {
+			if doErr != nil {
+				result = "error: " + doErr.Error()
+			}
+			if h.MaxFailures > 0 {
+				ct := uint(0)
+				err = tx.QueryRow(`SELECT count(*) FROM harmony_task_history 
+				WHERE task_id=$1 AND result=FALSE`, tID).Scan(&ct)
+				if err != nil {
+					logger.Error("Could not read task history:", err)
+					return false
+				}
+				if ct > h.MaxFailures {
+					_, err = tx.Exec("DELETE FROM harmony_task WHERE id=$1", tID)
+					if err != nil {
+						logger.Error("Could not delete failed job: ", err)
+						return false
+					}
+					// Note: Extra Info is left laying around for later review & clean-up
+				}
+			}
+		}
+		_, err = tx.Exec(`INSERT INTO harmony_task_history 
+												(task_id, name, posted,    work_start, work_end, result, err)
+			 VALUES ($1, $2, $3, $4, $5, $6, $7)`, tID, h.Name, postedTime, workStart, workEnd, done, result)
+		if err != nil {
+			logger.Error("Could not write history: ", err)
+			return false
+		}
+		return true
+	})
+	if err != nil {
+		logger.Error("Could not record transaction: ", err)
+		return
+	}
+	if !cm {
+		logger.Error("Committing the task records failed")
+	}
+}
+
+func (h *taskTypeHandler) AssertMachineHasCapacity() error {
+	r := h.TaskEngine.resourcesInUse()
+
+	if r.Cpu-h.Cost.Cpu < 0 {
+		return errors.New("Did not accept " + h.Name + " task: out of cpu")
+	}
+	if h.Cost.Ram > r.Ram {
+		return errors.New("Did not accept " + h.Name + " task: out of RAM")
+	}
+	if r.Gpu-h.Cost.Gpu < 0 {
+		return errors.New("Did not accept " + h.Name + " task: out of available GPU")
+	}
+	if h.Cost.Scratch > r.Scratch {
+		return errors.New("Did not accept " + h.Name + " task: out of scratch space")
```

Comment:
Scratch space is a bit more tricky, because:
* One worker may have multiple separate scratch space disks operating with independent partitions / mounts.
* Multiple workers may share one scratch space path.

The sector-storage subsystem has a mechanism to reserve storage when paths are shared between machines, that works through the sector index.

Here we could have a mechanism, where:
* We would check if we can claim some work
* If we can, we'd set up scratch space reservation, which for harmonytask would probably be a per-task-type callback like `ReserveResources(some-task-details-because-storage-reservations-need-sector-id) (free func(), err error)`
* If that doesn't succeed, we don't claim the task, and maybe set some backoff on trying to claim a task again
* After the work is done, we'd free the reservation.